File: proposed-projects/round-trip-experiment/reverse/20260909T024849Z_b96078c6/20260910T062159Z_d79086d3/ported_implementation/gbnf/grammar_graph/pointers.py
from __future__ import annotations

from typing import Dict, Iterator, ValuesView

from .graph_pointer import GraphPointer


class Pointers:
    def __init__(self, *pointers: GraphPointer) -> None:
        self._pointers: Dict[str, GraphPointer] = {}
        for pointer in pointers:
            self.add(pointer)

    def add(self, pointer: GraphPointer) -> None:
        self._pointers[pointer.id] = pointer

    def values(self) -> ValuesView[GraphPointer]:
        return self._pointers.values()

    def __iter__(self) -> Iterator[GraphPointer]:
        return iter(self._pointers.values())

    def __len__(self) -> int:
        return len(self._pointers)

    @property
    def size(self) -> int:
        return len(self._pointers)

    def __repr__(self) -> str:
        return f"<Pointers {', '.join(str(p.node.rule) for p in self)}>"

File: packages/gbnf-experiment/data/20260909T024849Z_b96078c6/ported_implementation/src/grammar-graph/pointers.ts
import type { GraphPointer } from "./graph-pointer.ts";

export class Pointers {
  #pointers = new Map<string, GraphPointer>();

  constructor(...pointers: GraphPointer[]) {
    for (const pointer of pointers) {
      this.add(pointer);
    }
  }

  add(pointer: GraphPointer): void {
    this.#pointers.set(pointer.id, pointer);
  }

  values(): IterableIterator<GraphPointer> {
    return this.#pointers.values();
  }

  [Symbol.iterator](): IterableIterator<GraphPointer> {
    return this.#pointers.values();
  }

  get size(): number {
    return this.#pointers.size;
  }

  toString(): string {
    return `<Pointers ${[...this].map((pointer) => pointer.node.rule).join(", ")}>`;
  }
}
